### PR TITLE
modifications for round

### LINKED ISSUE #12056: round for Float loses precision
The `__round__` method on `Float` uses `float`, which causes it to lose precision.

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

closes #12056

#### Brief description of what is fixed or changed

* The round method has been modified to round to even by default
in keeping with Python 3 semantics

* The round method no longer casts args to float so precision
is no longer lost.

* To get Python 3 behavior in Python 2, import round from sympy.core.compatibility

#### Other comments

Python and SymPy use different methods of rounding so rounding to the right of the decimal is WYSIWIG in SymPy but not in Python:
```python
>>> round(12.345,2)
12.35
>>> S(12.345).round(2)
12.34
```
This is because Python uses the full binary representation while SymPy rounds an integer based on the precision of the number, e.g. if a number `n` is stored as 0.3499999 with 2 digits of precision, SymPy would use 35 for the rounding since this is the closest integer to `n*10**2`.

cf [discussion here](https://groups.google.com/forum/#!topic/sympy/CgkxNTT85aI)

#### Release Notes

- core
  - round added to compatibility
  - under Python 3, Number.round == round(Number)
  - Number.round rounds to even on tie and does so in cases that Python misses, e.g. round(12.345,2) -> 12.35 in Python but 12.34 in SymPy whereas round(4.5) -> 4 in both.
  - Integer input is now returned as Integer

### CODE REVIEW

**Comment by smichr** [comment]
@asmeurer, do you want this in 1.4 or should it wait?

**Comment by asmeurer** [comment]
It's too late for 1.4 unless it is an absolutely critical bug fix. I plan to do the final release tomorrow or Wednesday.

**Comment by smichr** [comment]
> absolutely critical bug fix

Nope -- just creating a compatibilty version of round and changing default rounding semantics

**Comment by Sc0rpi0n101** [comment]
Some of the doctests are failing 
for eg, in pi.round()
it's Expected: 3
Got: 3. not 3.0 or 3
It's with the single number results only. All others are like 2.0*I or 6.0*3.0*I(still failing as int was expected)
is this a bug with Travis?

**Comment by smichr** [comment]
> Some of the doctests are failing

Thanks -- updated accordingly:
```
sympy\core\expr.py[33] .................................                   [OK]
```

**Comment by smichr** [comment]
It would be good to have another set of eyes check the changes. 

Does `how` work as a rounding method keyword? An alternative would be `rnd` but that looks like "random" to me...but that is the variables used in `evalf.py`

File: sympy/core/expr.py [hunk 1]
```diff
@@ -3406,12 +3419,23 @@ def round(self, p=0):
...
-            rv /= mag
-        rv = Rational(rv, q)
+        rv = Rational(rv, mag)
+        if how == 'even':
+            # rv >= self*mag
+            i = rv*mag
+            if 2*(i - self*mag) == 1 and i%2:
+                # round toward even
+                rv = Rational(i - 1, mag)
+        elif how != 'up':
```
Comment (asmeurer) [inline sympy/core/expr.py]: "up" is ambiguous. Does this mean towards +infinity or away from 0?

Comment (smichr) [reply sympy/core/expr.py]: I vote for not supporting custom rounding. That way the round function and method are synonymous in behavior.

Comment (smichr) [reply sympy/core/expr.py]: But to answer the question, it rounds away from 0.

File: sympy/core/expr.py [hunk 2]
```diff
@@ -3325,50 +3325,63 @@ def invert(self, g, *gens, **args):
             return mod_inverse(self, g)
         return invert(self, g, *gens, **args)
 
-    def round(self, p=0):
+    def round(self, p=0, how='even'):
```
Comment (asmeurer) [inline sympy/core/expr.py]: I would use a more descriptive name for the keyword, like `rounding_type`.

**Comment by asmeurer** [comment]
Some tests that `round(Float)` matches `round(float)` in Python 3 would be good.

File: sympy/core/expr.py (hunk 1 above)
Comment (asmeurer) [reply sympy/core/expr.py]: That's fine too. There are many rounding methods which could be added, but they don't need to be done in this PR.

**Comment by smichr** [comment]
I have to look at the irregular output I am getting.

**Comment by smichr** [comment]
I'm not sure what to do about rounding Integers. What should the following give and why?
```
>>> round(123)  # matches Python single-argument rules
123
>>> [round(S(123), i) for i in (-1,0,1)]  # preserves precision information
[1.2e+2, 123., 123.0]
```

**Comment by Sc0rpi0n101** [comment]
Shouldn't round(123,0) be 123, an int 
instead of 123. , a float with 0 decimal places?

**Comment by smichr** [comment]
See OP for how SymPy and Python differ in rounding algorithms.

@asmeurer , comparisons of types are now tested.

**Comment by smichr** [comment]
> Shouldn't round(123,0) be 123, an int

In Python 3, yes. This is now the result given by SymPy in this branch.

**Comment by smichr** [comment]
wip to make sure all SymPy numbers round to even on tie with given precision

**Comment by smichr** [comment]
wip to make sure all SymPy numbers round to even on tie with given precision

File: sympy/core/tests/test_expr.py [hunk 3]
```diff
@@ -1676,13 +1676,28 @@ def test_round():
...
+            n = '-' + n
+        v = str(S(n).round(p))[:j]  # pertinent digits
+        if v.endswith('.'):
+          continue  # it ends with 0 which is even
+        L = int(v[-1])  # last digit
+        assert L % 2 == 0, (n, '->', v)
 
     assert (Float(.3, 3) + 2*pi).round() == 7
     assert (Float(.3, 3) + 2*pi*100).round() == 629
-    assert (Float(.03, 3) + 2*pi/100).round(5) == 0.09283
```
Comment (smichr) [inline sympy/core/tests/test_expr.py]: the output is limited by the precision of 3

**Comment by smichr** [comment]
I feel like the test that were in place and those that I have added are a good test of this algorithm. I am also happy that we have round-to-even rounding working for Floats. The simplified code here belies the difficulty of getting this right. If anyone else wants to give this a check that would be nice, otherwise I will commit this post-Easter.

**Comment by oscarbenjamin** [comment]
+1

Should a new issue be opened with the dropping Python 2 label noting to follow up on this later?

**Comment by smichr** [comment]
cf #16686

**Comment by cbm755** [comment]
I'd like to (eventually) see Float rounding done by mpmath (https://github.com/fredrik-johansson/mpmath/issues/455)

Then we could worry separately about how Rationals, irrational Numbers, etc are dealt with, without the code for Floats being interspersed.  But that could wait until Python 2 is dropped, so I'm ok in principle with some of these changes.

- - - -

But the round-to-even bit is suspicious for Floats.  I don't believe the "augment precision, add 5 to previously least significant digit" logic.

Here is a float, converted to a Float, that is less than 575/1000 but none-the-less rounds (incorrectly) to 0.58:
```
>>> a = Float(0.575 - 6e-16)
>>> a < Rational(575, 1000)
True
>>> round(a, 2)
0.58
```

**Comment by smichr** [comment]
@cbm77, please see inline notes to the algorithm which is now recast more explicitly in terms of known digits as an integer and the rounding digit in the 10ths place.

**Pull request merged**